File: node-service/tests/unit/calculatorService.test.js
/**
 * Jest unit tests for CalculatorService
 */

const CalculatorService = require('../../src/calculatorService');

describe('CalculatorService', () => {
  let calculator;

  beforeEach(() => {
    calculator = new CalculatorService();
  });

  describe('Addition', () => {
    test('should add two positive numbers correctly', () => {
      const result = calculator.add(5, 3);
      expect(result).toBe(8);
    });

    test('should add two positive numbers correctly - INTENTIONALLY FAILING', () => {
      const result = calculator.add(5, 3);
      expect(result).toBe(10); // This will fail - should be 8
    });

    test('should add negative numbers correctly', () => {
      const result = calculator.add(-5, -3);
      expect(result).toBe(-8);
    });

    test('should add zero correctly', () => {
      const result = calculator.add(5, 0);
      expect(result).toBe(5);
    });

    test.each([
      [1, 2, 3],
      [0.5, 0.3, 0.8],
      [-1, 1, 0],
      [100, 200, 300],
      [0, 0, 0]
    ])('should add %i and %i to get %i', (a, b, expected) => {
      const result = calculator.add(a, b);
      expect(result).toBe(expected);
    });
  });

  describe('Subtraction', () => {
    test('should subtract two positive numbers correctly', () => {
      const result = calculator.subtract(5, 3);
      expect(result).toBe(2);
    });

    test('should handle negative result', () => {
      const result = calculator.subtract(3, 5);
      expect(result).toBe(-2);
    });

    test.each([
      [5, 3, 2],
      [0, 0, 0],
      [10, 5, 5],
      [-5, -3, -2]
    ])('should subtract %i from %i to get %i', (a, b, expected) => {
      const result = calculator.subtract(a, b);
      expect(result).toBe(expected);
    });
  });

  describe('Multiplication', () => {
    test('should multiply two positive numbers correctly', () => {
      const result = calculator.multiply(5, 3);
      expect(result).toBe(15);
    });

    test('should handle multiplication by zero', () => {
      const result = calculator.multiply(5, 0);
      expect(result).toBe(0);
    });

    test.each([
      [2, 3, 6],
      [0.5, 0.5, 0.25],
      [-2, 3, -6],
      [0, 100, 0]
    ])('should multiply %i by %i to get %i', (a, b, expected) => {
      const result = calculator.multiply(a, b);
      expect(result).toBe(expected);
    });
  });

  describe('Division', () => {
    test('should divide two positive numbers correctly', () => {
      const result = calculator.divide(6, 2);
      expect(result).toBe(3);
    });

    test('should throw error when dividing by zero', () => {
      expect(() => {
        calculator.divide(5, 0);
      }).toThrow('Division by zero is not allowed');
    });

    test.each([
      [6, 2, 3],
      [5, 2, 2.5],
      [0, 5, 0],
      [-6, 2, -3]
    ])('should divide %i by %i to get %i', (a, b, expected) => {
      const result = calculator.divide(a, b);
      expect(result).toBe(expected);
    });
  });

  describe('Power', () => {
    test('should calculate power correctly', () => {
      const result = calculator.power(2, 3);
      expect(result).toBe(8);
    });

    test('should handle power of zero', () => {
      const result = calculator.power(5, 0);
      expect(result).toBe(1);
    });

    test.each([
      [2, 3, 8],
      [5, 0, 1],
      [2, -1, 0.5],
      [0, 5, 0]
    ])('should calculate %i to the power of %i to get %i', (base, exponent, expected) => {
      const result = calculator.power(base, exponent);
      expect(result).toBe(expected);
    });
  });

  describe('Square Root', () => {
    test('should calculate square root correctly', () => {
      const result = calculator.sqrt(16);
      expect(result).toBe(4);
    });

    test('should throw error for negative number', () => {
      expect(() => {
        calculator.sqrt(-4);
      }).toThrow('Cannot calculate square root of negative number');
    });

    test.each([
      [0, 0],
      [1, 1],
      [4, 2],
      [9, 3],
      [16, 4],
      [25, 5]
    ])('should calculate square root of %i to get %i', (number, expected) => {
      const result = calculator.sqrt(number);
      expect(result).toBe(expected);
    });
  });

  describe('Factorial', () => {
    test('should calculate factorial correctly', () => {
      const result = calculator.factorial(5);
      expect(result).toBe(120);
    });

    test('should handle factorial of zero', () => {
      const result = calculator.factorial(0);
      expect(result).toBe(1);
    });

    test('should handle factorial of one', () => {
      const result = calculator.factorial(1);
      expect(result).toBe(1);
    });

    test('should throw error for negative number', () => {
      expect(() => {
        calculator.factorial(-5);
      }).toThrow('Factorial is not defined for negative numbers');
    });

    test.each([
      [0, 1],
      [1, 1],
      [2, 2],
      [3, 6],
      [4, 24],
      [5, 120]
    ])('should calculate factorial of %i to get %i', (n, expected) => {
      const result = calculator.factorial(n);
      expect(result).toBe(expected);
    });
  });

  describe('Array Operations', () => {
    describe('Average', () => {
      test('should calculate average correctly', () => {
        const result = calculator.average([1, 2, 3, 4, 5]);
        expect(result).toBe(3);
      });

      test('should throw error for empty array', () => {
        expect(() => {
          calculator.average([]);
        }).toThrow('Cannot calculate average of empty array');
      });

      test('should throw error for null array', () => {
        expect(() => {
          calculator.average(null);
        }).toThrow('Cannot calculate average of empty array');
      });
    });

    describe('Maximum', () => {
      test('should find maximum correctly', () => {
        const result = calculator.max([1, 5, 3, 9, 2]);
        expect(result).toBe(9);
      });

      test('should throw error for empty array', () => {
        expect(() => {
          calculator.max([]);
        }).toThrow('Cannot find maximum of empty array');
      });

      test('should throw error for null array', () => {
        expect(() => {
          calculator.max(null);
        }).toThrow('Cannot find maximum of empty array');
      });
    });

    describe('Minimum', () => {
      test('should find minimum correctly', () => {
        const result = calculator.min([1, 5, 3, 9, 2]);
        expect(result).toBe(1);
      });

      test('should throw error for empty array', () => {
        expect(() => {
          calculator.min([]);
        }).toThrow('Cannot find minimum of empty array');
      });

      test('should throw error for null array', () => {
        expect(() => {
          calculator.min(null);
        }).toThrow('Cannot find minimum of empty array');
      });
    });
  });

  describe('GCD', () => {
    test('should calculate GCD of positive numbers', () => {
      const result = calculator.gcd(48, 18);
      expect(result).toBe(6);
    });

    test('should handle GCD with zero', () => {
      const result = calculator.gcd(0, 5);
      expect(result).toBe(5);
    });

    test.each([
      [48, 18, 6],
      [54, 24, 6],
      [7, 13, 1],
      [0, 0, 0]
    ])('should calculate GCD of %i and %i to get %i', (a, b, expected) => {
      const result = calculator.gcd(a, b);
      expect(result).toBe(expected);
    });
  });

  describe('LCM', () => {
    test('should calculate LCM of positive numbers', () => {
      const result = calculator.lcm(12, 18);
      expect(result).toBe(36);
    });

    test('should handle LCM with zero', () => {
      const result = calculator.lcm(0, 5);
      expect(result).toBe(0);
    });

    test.each([
      [12, 18, 36],
      [8, 12, 24],
      [7, 13, 91],
      [0, 0, 0]
    ])('should calculate LCM of %i and %i to get %i', (a, b, expected) => {
      const result = calculator.lcm(a, b);
      expect(result).toBe(expected);
    });
  });

  describe('Prime Numbers', () => {
    test('should identify prime numbers correctly', () => {
      expect(calculator.isPrime(2)).toBe(true);
      expect(calculator.isPrime(3)).toBe(true);
      expect(calculator.isPrime(17)).toBe(true);
    });

    test('should identify non-prime numbers correctly', () => {
      expect(calculator.isPrime(1)).toBe(false);
      expect(calculator.isPrime(4)).toBe(false);
      expect(calculator.isPrime(15)).toBe(false);
    });

    test.each([
      [2, true],
      [3, true],
      [4, false],
      [5, true],
      [6, false],
      [7, true],
      [8, false],
      [9, false],
      [10, false],
      [11, true]
    ])('should identify %i as prime: %s', (n, expected) => {
      const result = calculator.isPrime(n);
      expect(result).toBe(expected);
    });
  });
});

File: java-service/src/test/java/com/example/unit/CalculatorServiceTest.java
package com.example.unit;

import com.example.CalculatorService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("CalculatorService Unit Tests")
class CalculatorServiceTest {

    private CalculatorService calculatorService;

    @BeforeEach
    void setUp() {
        calculatorService = new CalculatorService();
    }

    @Nested
    @DisplayName("Addition Tests")
    class AdditionTests {

        @Test
        @DisplayName("Should add two positive numbers correctly")
        void shouldAddTwoPositiveNumbers() {
            // Given
            double a = 5.0;
            double b = 3.0;

            // When
            double result = calculatorService.add(a, b);

            // Then
            assertEquals(8.0, result, 0.001);
        }

        @Test
        @DisplayName("Should add negative numbers correctly")
        void shouldAddNegativeNumbers() {
            // Given
            double a = -5.0;
            double b = -3.0;

            // When
            double result = calculatorService.add(a, b);

            // Then
            assertEquals(-8.0, result, 0.001);
        }

        @Test
        @DisplayName("Should add zero correctly")
        void shouldAddZero() {
            // Given
            double a = 5.0;
            double b = 0.0;

            // When
            double result = calculatorService.add(a, b);

            // Then
            assertEquals(5.0, result, 0.001);
        }

        @ParameterizedTest
        @CsvSource({
            "1.0, 2.0, 3.0",
            "0.5, 0.3, 0.8",
            "-1.0, 1.0, 0.0",
            "100.0, 200.0, 300.0"
        })
        @DisplayName("Should add various number combinations")
        void shouldAddVariousNumbers(double a, double b, double expected) {
            // When
            double result = calculatorService.add(a, b);

            // Then
            assertEquals(expected, result, 0.001);
        }
    }

    @Nested
    @DisplayName("Subtraction Tests")
    class SubtractionTests {

        @Test
        @DisplayName("Should subtract two positive numbers correctly")
        void shouldSubtractTwoPositiveNumbers() {
            // Given
            double a = 5.0;
            double b = 3.0;

            // When
            double result = calculatorService.subtract(a, b);

            // Then
            assertEquals(2.0, result, 0.001);
        }

        @Test
        @DisplayName("Should handle negative result")
        void shouldHandleNegativeResult() {
            // Given
            double a = 3.0;
            double b = 5.0;

            // When
            double result = calculatorService.subtract(a, b);

            // Then
            assertEquals(-2.0, result, 0.001);
        }

        @ParameterizedTest
        @CsvSource({
            "5.0, 3.0, 2.0",
            "0.0, 0.0, 0.0",
            "10.0, 5.0, 5.0",
            "-5.0, -3.0, -2.0"
        })
        @DisplayName("Should subtract various number combinations")
        void shouldSubtractVariousNumbers(double a, double b, double expected) {
            // When
            double result = calculatorService.subtract(a, b);

            // Then
            assertEquals(expected, result, 0.001);
        }
    }

    @Nested
    @DisplayName("Multiplication Tests")
    class MultiplicationTests {

        @Test
        @DisplayName("Should multiply two positive numbers correctly")
        void shouldMultiplyTwoPositiveNumbers() {
            // Given
            double a = 5.0;
            double b = 3.0;

            // When
            double result = calculatorService.multiply(a, b);

            // Then
            assertEquals(15.0, result, 0.001);
        }

        @Test
        @DisplayName("Should handle multiplication by zero")
        void shouldHandleMultiplicationByZero() {
            // Given
            double a = 5.0;
            double b = 0.0;

            // When
            double result = calculatorService.multiply(a, b);

            // Then
            assertEquals(0.0, result, 0.001);
        }

        @ParameterizedTest
        @CsvSource({
            "2.0, 3.0, 6.0",
            "0.5, 0.5, 0.25",
            "-2.0, 3.0, -6.0",
            "0.0, 100.0, 0.0"
        })
        @DisplayName("Should multiply various number combinations")
        void shouldMultiplyVariousNumbers(double a, double b, double expected) {
            // When
            double result = calculatorService.multiply(a, b);

            // Then
            assertEquals(expected, result, 0.001);
        }
    }

    @Nested
    @DisplayName("Division Tests")
    class DivisionTests {

        @Test
        @DisplayName("Should divide two positive numbers correctly")
        void shouldDivideTwoPositiveNumbers() {
            // Given
            double a = 6.0;
            double b = 2.0;

            // When
            double result = calculatorService.divide(a, b);

            // Then
            assertEquals(3.0, result, 0.001);
        }

        @Test
        @DisplayName("Should throw exception when dividing by zero")
        void shouldThrowExceptionWhenDividingByZero() {
            // Given
            double a = 5.0;
            double b = 0.0;

            // When & Then
            ArithmeticException exception = assertThrows(ArithmeticException.class, () -> {
                calculatorService.divide(a, b);
            });

            assertEquals("Division by zero is not allowed", exception.getMessage());
        }

        @ParameterizedTest
        @CsvSource({
            "6.0, 2.0, 3.0",
            "5.0, 2.0, 2.5",
            "0.0, 5.0, 0.0",
            "-6.0, 2.0, -3.0"
        })
        @DisplayName("Should divide various number combinations")
        void shouldDivideVariousNumbers(double a, double b, double expected) {
            // When
            double result = calculatorService.divide(a, b);

            // Then
            assertEquals(expected, result, 0.001);
        }
    }

    @Nested
    @DisplayName("Power Tests")
    class PowerTests {

        @Test
        @DisplayName("Should calculate power correctly")
        void shouldCalculatePowerCorrectly() {
            // Given
            double base = 2.0;
            double exponent = 3.0;

            // When
            double result = calculatorService.power(base, exponent);

            // Then
            assertEquals(8.0, result, 0.001);
        }

        @Test
        @DisplayName("Should handle power of zero")
        void shouldHandlePowerOfZero() {
            // Given
            double base = 5.0;
            double exponent = 0.0;

            // When
            double result = calculatorService.power(base, exponent);

            // Then
            assertEquals(1.0, result, 0.001);
        }

        @ParameterizedTest
        @CsvSource({
            "2.0, 3.0, 8.0",
            "5.0, 0.0, 1.0",
            "2.0, -1.0, 0.5",
            "0.0, 5.0, 0.0"
        })
        @DisplayName("Should calculate various power combinations")
        void shouldCalculateVariousPowerCombinations(double base, double exponent, double expected) {
            // When
            double result = calculatorService.power(base, exponent);

            // Then
            assertEquals(expected, result, 0.001);
        }
    }

    @Nested
    @DisplayName("Square Root Tests")
    class SquareRootTests {

        @Test
        @DisplayName("Should calculate square root correctly")
        void shouldCalculateSquareRootCorrectly() {
            // Given
            double number = 16.0;

            // When
            double result = calculatorService.sqrt(number);

            // Then
            assertEquals(4.0, result, 0.001);
        }

        @Test
        @DisplayName("Should throw exception for negative number")
        void shouldThrowExceptionForNegativeNumber() {
            // Given
            double number = -4.0;

            // When & Then
            IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> {
                calculatorService.sqrt(number);
            });

            assertEquals("Cannot calculate square root of negative number", exception.getMessage());
        }

        @ParameterizedTest
        @ValueSource(doubles = {0.0, 1.0, 4.0, 9.0, 16.0, 25.0})
        @DisplayName("Should calculate square root of perfect squares")
        void shouldCalculateSquareRootOfPerfectSquares(double number) {
            // When
            double result = calculatorService.sqrt(number);

            // Then
            double expected = Math.sqrt(number);
            assertEquals(expected, result, 0.001);
        }
    }

    @Nested
    @DisplayName("Factorial Tests")
    class FactorialTests {

        @Test
        @DisplayName("Should calculate factorial correctly")
        void shouldCalculateFactorialCorrectly() {
            // Given
            int n = 5;

            // When
            long result = calculatorService.factorial(n);

            // Then
            assertEquals(120L, result);
        }

        @Test
        @DisplayName("Should handle factorial of zero")
        void shouldHandleFactorialOfZero() {
            // Given
            int n = 0;

            // When
            long result = calculatorService.factorial(n);

            // Then
            assertEquals(1L, result);
        }

        @Test
        @DisplayName("Should handle factorial of one")
        void shouldHandleFactorialOfOne() {
            // Given
            int n = 1;

            // When
            long result = calculatorService.factorial(n);

            // Then
            assertEquals(1L, result);
        }

        @Test
        @DisplayName("Should throw exception for negative number")
        void shouldThrowExceptionForNegativeNumber() {
            // Given
            int n = -5;

            // When & Then
            IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> {
                calculatorService.factorial(n);
            });

            assertEquals("Factorial is not defined for negative numbers", exception.getMessage());
        }

        @ParameterizedTest
        @CsvSource({
            "0, 1",
            "1, 1",
            "2, 2",
            "3, 6",
            "4, 24",
            "5, 120"
        })
        @DisplayName("Should calculate factorial for various numbers")
        void shouldCalculateFactorialForVariousNumbers(int n, long expected) {
            // When
            long result = calculatorService.factorial(n);

            // Then
            assertEquals(expected, result);
        }
    }

    @Nested
    @DisplayName("Array Operations Tests")
    class ArrayOperationsTests {

        @Test
        @DisplayName("Should calculate average correctly")
        void shouldCalculateAverageCorrectly() {
            // Given
            double[] numbers = {1.0, 2.0, 3.0, 4.0, 5.0};

            // When
            double result = calculatorService.average(numbers);

            // Then
            assertEquals(3.0, result, 0.001);
        }

        @Test
        @DisplayName("Should throw exception for empty array average")
        void shouldThrowExceptionForEmptyArrayAverage() {
            // Given
            double[] numbers = {};

            // When & Then
            IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> {
                calculatorService.average(numbers);
            });

            assertEquals("Cannot calculate average of empty array", exception.getMessage());
        }

        @Test
        @DisplayName("Should find maximum correctly")
        void shouldFindMaximumCorrectly() {
            // Given
            double[] numbers = {1.0, 5.0, 3.0, 9.0, 2.0};

            // When
            double result = calculatorService.max(numbers);

            // Then
            assertEquals(9.0, result, 0.001);
        }

        @Test
        @DisplayName("Should find minimum correctly")
        void shouldFindMinimumCorrectly() {
            // Given
            double[] numbers = {1.0, 5.0, 3.0, 9.0, 2.0};

            // When
            double result = calculatorService.min(numbers);

            // Then
            assertEquals(1.0, result, 0.001);
        }
    }
}
